synfig-studio/src/gui/duckmatic.cpp

```diff
@@ -815,36 +815,26 @@ Duckmatic::set_guides_color(const synfig::Color &c)
 		signal_grid_changed().emit();
 	}
 }
-
-Duckmatic::GuideList::iterator
-Duckmatic::find_guide_x(synfig::Point pos, float radius)
+double
+Duckmatic::calculate_distance_from_guide(const Guide& guide, const Point& point)
 {
-	GuideList::iterator iter,best(guide_list_x_.end());
-	float dist(radius);
-	for(iter=guide_list_x_.begin();iter!=guide_list_x_.end();++iter)
-	{
-		float amount(std::fabs(*iter-pos[0]));
-		if(amount<dist)
-		{
-			dist=amount;
-			best=iter;
-		}
-	}
-	return best;
+	return std::fabs((cos(guide.angle.get())*(guide.point[1]-point[1])) - (sin(guide.angle.get())*(guide.point[0] - point[0])));
 }
 
+
 Duckmatic::GuideList::iterator
-Duckmatic::find_guide_y(synfig::Point pos, float radius)
+Duckmatic::find_guide(synfig::Point pos, float radius, Guide* second_best_guide_match)
```

Comment:
The last argument does not work as intended.
As it is an optional argument and it stores a pointer, you must use `Guide** second_best_guide_match` AND set it in this method something like `*second_best_guide_match = nullptr` (and check if it is not `nullptr` first).

Reply:
In other words: you are passong the pointer as value, but you pass ot as reference

Reply:
> The last argument does not work as intended. As it is an optional argument and it stores a pointer, you must use `Guide** second_best_guide_match` AND set it in this method something like `*second_best_guide_match = nullptr` (and check if it is not `nullptr` first).

This turns out to be a harder task than i anticipated xD. So what I wanted to do here is to pass this pointer and set its value to be null inside the method and then if it is found it becomes not null. this is how i know in the method `snap_point_to_grid` if there is a second best match or not (by checking to see if its not null after it was sent to find_guide). But then if its null I cant set it to point to the iterator without causing a seg error.

Another alternative I tried quickly is to instead switch over to using iterators and setting it to the end of our guide list instead of null but since `snap_point_to_grid` is const its not possible to set the `second_best_guide_match` iterator to be that.

or maybe it is that i'm sleepy rn and missing something o_O.

Reply:
You can use iterators.

But if you want to understand: https://www.geeksforgeeks.org/passing-reference-to-a-pointer-in-c/

Reply:
> You can use iterators.

but then the issue as I mentioned was that I would'nt be able to inialize the iterator to be `guide_list_.end()` in the `snap_point_to_grid` method as it is const.

> But if you want to understand: https://www.geeksforgeeks.org/passing-reference-to-a-pointer-in-c/

I also tried having the argument be a refrence to a pointer but then it cant be null. It seems the only reasonable option here is using a pointer to a pointer, but this is still causing some issues.

so basically the pointer i am sending is null so the argument is now pointing to null and when i try to access the pointer which is being sent I cant because I would be accesing a null pointer. a bit of a closed loop lol.

Reply:
> I would'nt be able to inialize the iterator to be guide_list_.end()

Try `cend`.


And read the text and my first comment again :)

Reply:
> 
> And read the text and my first comment again if you want to use pointers :)

oh yes my bad, I realized what was causing the confusion for me xD. I needed to check if the arg itself was not a nullptr not what it points to.

Reply:
Ok, I believe now I found the other issue with snapping to intersection of rotated guides. It still doesn't work "perfectly" though but at least it works like the previous implementation.

What I mean is that if there is no second best guide match there would be no snapping to intersection. second best guide match is present if there is a guide found (distance less than a pre-specified value `radius` in `find_guide`) other than the main current guide.

I will mark this as resolved now but if you feel like there is still an issue feel free to reopen this :)